### PR TITLE
Add argument asserts to copyForwards and copyBackwards

### PR DESCRIPTION
**Pull request by moosichu**
Helps document in code what the comments already say. I had to fixup a comptime use case of these functions in Encoding.zig (as you can't retrieve the pointer value at comptime...). However, @memcpy seems to work at comptime now so it resolves a pre-existing TODO there anyway!

It does raise the question of how you should handle comptime memcpys with overlapping arrays though - as ideally you would be able to still validate the ptr ordering in that case as well?

This also fixes a bug where an assert in `copyBackwards` was effectively doing nothing due to an `@setRuntimeSafety` call in the same scope.

### CODE REVIEW

**Comment by moosichu** [comment]
I mainly made this because I got the intended uses the wrong way around 🤦

**Comment by moosichu** [comment]
Obviously I completed a TODO but didn't remove it 🤦  (that is now fixed!)

**Comment by rohlem** [comment]
You can put this check behind an `if(!@inComptime())` to make it runtime-only. That way `comptime` calls won't be affected.

There is another issue however in that two non-overlapping slices (say from disconnected allocations) should also be free to use either version (even if `@memcpy` would be more optimal for that case - maybe the same code path sometimes overlaps in the correct direction and sometimes doesn't overlap at all).
So the correct check IMO would be to only assert the ordering if the slices actually do overlap.

<details><summary>code simplified from my own codebase for reference:</summary>
<p>

```zig
fn copyToBeforeSelfOrNoOverlap(self_source: anytype, destination: anytype) void {
    if (std.debug.runtime_safety and !@inComptime()) {
        const source_begin_address = @intFromPointer(self_source.ptr);
        const source_end_address = source_begin_address + self_source.len;
        const destination_begin_address = @intFromPointer(destination.ptr);
        if ((destination_begin_address > source_begin_address) and (destination_begin_address < source_end_address)) {
            panic("invalid use of copyToBeforeSelfOrNoOverlap detected: destination lies after source; should have used copyToAfterSelfOrNoOverlap!\ntried copying from {} to {}", .{ self_source, destination });
        }
    }
    std.mem.copyForwards(Element, destination, self_source); //old version before this PR
}
fn copyToAfterSelfOrNoOverlap(self_source: anytype, destination: anytype) void {
    if (std.debug.runtime_safety and !@inComptime()) {
        const source_begin_address = @intFromPtr(self_source.ptr);
        const destination_begin_address = @intFromPtr(destination.ptr);
        const destination_end_address = destination_begin_address + destination.len;
        const destination_end_used_address = destination_begin_address + self_source.len;
        if ((source_begin_address > destination_begin_address) and (source_begin_address < destination_end_address)) {
            if (source_begin_address > destination_end_used_address) {
                std.log.warn("potentially invalid use of copyToAfterSelfOrNoOverlap detected: destination lies before source (but dynamically ok because of source length); should probably have used copyToBeforeSelfOrNoOverlap!\ncopying from {} to {}", .{ self_source, destination });
                std.debug.dumpCurrentStackTrace(null);
            } else {
                panic("invalid use of copyToAfterSelfOrNoOverlap detected: destination lies before source; should have used copyToBeforeSelfOrNoOverlap!\ntried copying from {} to {}", .{ self_source, destination });
            }
        }
    }
    std.mem.copyBackwards(Element, destination, self_source);
}
```
</p>
</details>

**Comment by moosichu** [comment]
I'm just testing that more comprehensive solution now...

However - something I haven't added to the PR but easily could is this as well using the building blocks this PR now contains:
```zig
/// Copy all or source into dest at position 0.
/// dest.len must be >= source.len.
/// Deals with overlapping/aliasing slices forwards or backwards.
/// Note: Cannot be called at comptime as there is currently no mechanism for
/// determining how comptime slices overlap.
pub fn copySafe(comptime T: type, dest: []T, source: []const T) void {
    if (@inComptime()) {
        @compileError("copySafe cannot currently be called at comptime, as there is no mechanism for determining if two slices alias each other available");
    }
    assert(dest.len >= source.len);
    if (!slicesOverlap(T, dest, source)) {
        @memcpy(dest[0..source.len], source);
        return;
    }
    if (@intFromPtr(source.ptr) >= @intFromPtr(dest.ptr)) {
        copyForwards(T, dest, source);
    } else {
        copyBackwards(T, dest, source);
    }
}
```

**Comment by rohlem** [comment]
@moosichu regarding `copySafe`, so far I don't have any callsites in my own code that would benefit from this mechanism.
That said, it's possible legitimate use cases exist, also it might be an easier option for newcomers / less performance-oriented code.
I'd vote for having it (should do no harm), but ultimately up to the maintainers to decide.

Regarding the implementation, I'm not sure about the branch deciding between `@memcpy` and `copyForward`:
 - My gut feeling says that for plain `copyForwards` and `copyBackwards` the compiler "should be smart enough" to understand the pattern, meaning an extra branch to plain `@memcpy` shouldn't improve performance.
 - If that is true, it probably won't help performance in `copySafe` either.

But that's also a decision for maintainers to weigh in, maybe some backends would benefit from a builtin `@safeMemcpy`?
Either way, having a concrete/realistic usage scenario as motivation would probably help the discussion.

File: lib/std/mem.zig
Comment (rohlem) [inline lib/std/mem.zig:214]: The new `copyForwards`, `copyBackwards` enforce pointer ordering if we return `true` here.
Since there's only danger of overwriting if `a_end != b_begin`, I think the `return`-ed condition should be `<` instead of `<=` - I think that's the source of the current CI errors.

**Comment by moosichu** [comment]
Updated the PR to fix some logic and clarify what is going on in the implementation.

File: lib/std/mem.zig
Comment (moosichu) [reply lib/std/mem.zig]: yeah! independetntly noticed that when reviewing things myself, cheers for that! 👍

**Comment by rohlem** [comment]
I realized there's still optimization potential:
Because each call site really only cares about one-directional overlap, instead of a bi-directional `sliceOverlap` check,
a more optimal implementation would checking `if (sliceContainsPointer(a, b.ptr)) unreachable;`.
That effectively reduces the number of if-s on the hot path from 2 to 1.

**Comment by moosichu** [comment]
yeah... I considered something like that... but was really struggling to come up with a good, descriptive name for it.

That is a really neat one though so I will put something together tomorrow and check that in!

File: lib/std/mem.zig
Comment (andrewrk) [inline lib/std/mem.zig:233]:
```suggestion
```

This is already asserted by the for loop below.

File: src/arch/x86_64/Encoding.zig
Comment (andrewrk) [inline src/arch/x86_64/Encoding.zig:797]: Thank you for fixing these :+1:

File: lib/std/mem.zig
Comment (andrewrk) [inline lib/std/mem.zig:201]: Everything in std.mem already has to do with slices:

```suggestion
pub fn alias(comptime T: type, slice_a: []const T, slice_b: []const T) bool {
```

File: lib/std/mem.zig
Comment (andrewrk) [inline lib/std/mem.zig:226]: I find all of these comments and logic confusing, when you could do simply instead:

```suggestion
    return !(a_begin >= b_end or b_begin >= a_end);
```

**Review by andrewrk — CHANGES_REQUESTED** [review]
In addition to the problems identified in this review, the entire idea of checking whether aliasing is occurring in userland is flawed. You already noticed that it cannot be done at compile time, but there is more. This would wreck pointer provenance-based optimizations in ReleaseSafe mode because of the `@intFromPtr` calls.

I think general-purpose alias checking needs to be done via a language builtin instead.

File: lib/std/mem.zig
Comment (moosichu) [reply lib/std/mem.zig]: 🙈  I can overcomplicate simple things sometimes! That makes so much more sense!

**Comment by moosichu** [comment]
> I realized there's still optimization potential: Because each call site really only cares about one-directional overlap, instead of a bi-directional `slicesOverlap` check, a more optimal implementation would be checking `if (sliceContains(a, b.ptr)) unreachable;`. That effectively reduces the number of if-s on the hot path from 2 to 1.

I've decided not to do this... based on Andrew's comments if an @alias builtin will be introduced it makes sense to have a user-space implementation of that for now! (as that is what will have to be used once the builtin is introduced).

> In addition to the problems identified in this review, the entire idea of checking whether aliasing is occurring in userland is flawed. You already noticed that it cannot be done at compile time, but there is more. This would wreck pointer provenance-based optimizations in ReleaseSafe mode because of the `@intFromPtr` calls.
> 
> I think general-purpose alias checking needs to be done via a language builtin instead.

Based on this I hope you don't mind that I've gone ahead and added #17613. I can leave this PR open for now.

I personally think this might still be worth merging in the short term for the safety benefits it would bring (even if not ideal) and then fix it once the builtin is introduced. But otherwise it can be left open and modified once someone has done #17613. (I want to focus on the archiver for now... so won't touch it in the short-term at least).

**Comment by rohlem** [comment]
> You already noticed that it cannot be done at compile time, but there is more.

I do think that checking whether one pointer points within a still-modifiable slice of a `comptime` memory isle should be implementable as a compiler builtin.
With `comptime` pointers-to-mutable, the compiler already has to allow different pointers (with the same value) to modify the same location, so it's already required to track identity.
That is exactly what this use case would require, every scenario outside of that can gracefully result in a compile error.

I also think that the name `alias` is just as fitting for arguments `(slice, pointer)` as it is for arguments `(slice, slice)`.
The second argument can be made `anytype` and support both use cases (`anytype` parameters similarly exist in other `std.mem` functions as well).

Just my thoughts; feel free to let me know if I'm wrong / you disagree with any of the above course.
Otherwise I'll probably try implementing it myself (though not in the immediate future, more likely in a year or so).

**Comment by moosichu** [comment]
> I also think that the name alias is just as fitting for arguments (slice, pointer) as it is for arguments (slice, slice).
The second argument can be made anytype and support both use cases (anytype parameters similarly exist in other std.mem functions as well).

I've implemented this in a follow-up commit. Very happy to revert it based on feedback also, but thought it would be worth having for comparison's sake.

**Comment by moosichu** [comment]
No! You aren't being pedantic at all. Rushed this a little this morning before heading out and forgot about that bit so will do that later.

**Comment by moosichu** [comment]
saw the tests fail.. theres obviously some failure of logic but need to head to bed now and will look properly tomorrow!

**Comment by rohlem** [comment]
Well, I'm embarassed, the CI errors trigger because the new formulation doesn't work for a no-op copy where `source.ptr == destination.ptr`. Sorry for that oversight.
The optimal condition should be `ptr_int < range_end and ptr_int > range_begin`, but I'm not sure what a good name for the function would then be. `aliasDifferent`, `aliasAtOffset`?
We could also write the asserted condition as `!alias(...) or (source.ptr == dest.ptr)` and hope the optimizer merges them after inlining (though it probably wouldn't in `Debug` builds).

**Comment by moosichu** [comment]
I've gone for the:
`!alias(...) or (source.ptr == dest.ptr)`
approach without inlining, mainly because of the presumption is that `@alias` will eventually be a builtin with the API I've implemented, so that will be the building block the function will eventually have to use anyway.

**Pull request closed without merging**
